doc/modules/ROOT/pages/design/middleware.adoc

```diff
@@ -359,3 +359,21 @@ See link:../building_clients.html[Building Clients] on how to implement the serv
 Once the sideloader has been started, it can be triggered by an ordinary `eval` or `load-file` operation. The nREPL server will first try to find a resource/class on the classpath of its own JVM. Failing this, it will attempt to request it from the nREPL client by the above described mechanism. Should this still fail, expect a `ClassNotFoundException` as usual.
 
 TIP: Once a class has been loaded, it will become available to all sessions on nREPL server.
+
+== Dynamic middleware loading
+
+NOTE: Sideloading support was added in nREPL 0.8 and the API is still
+considered experimental, and may change.
+
+nREPL includes a `dynamic-loader` middleware, which can be used, at runtime,
+query and change the middleware stack the nREPL server is using. This is especially
+powerful when combined with sideloading, as it allows a client to configure the 
+server after connecting, and provides an alternative to having to specify the middleware required by the client as startup time.
+
+This introduces three new operations:
+
+* `ls-middleware`, to return a list of active middleware, ordered from inside outwards.
+* `add-middleware`, which adds a middleware to the stack. Optionally, a list of `extra-namespaces` could be provided for loading. This is useful when adding middleware that implement some form of deferred loading. Examples include `cider-nrepl` and `refactor-nrepl`. In these cases, some of the required namespaces might only be loaded upon first use, which may occur outside of a sideloading session, and thus fail. This feature allows us to pre-load namespaces when we add a middleware. If loading of any particular middleware fails, the stack will be unchanged.
+* `swap-middleware`, similar to `add-`, but replaces all existing middleware. Note that this _may_ remove the `dynamic-loader` itself.
```

Comment:
I couple of examples of using `add/swap`-middleware would be useful I guess. I think that using `swap-middleware` will be particularly tricky for some people. I'm not sure if this should go here or in the building clients section.

Reply:
Added section in "building clients"